File: src/main/java/org/zkoss/typeahead/Typeahead.java
package org.zkoss.typeahead;

import org.zkoss.typeahead.data.Dataset;
import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.sys.ComponentCtrl;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.Textbox;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;

/**
 * A <a href="https://github.com/corejavascript/typeahead.js">Typeahead.js</a> component.
 *
 * @author Sean Connolly
 */
public class Typeahead extends Textbox {

    static {
        addClientEvent(Typeahead.class, Events.ON_SELECT, ComponentCtrl.CE_DUPLICATE_IGNORE);
        addClientEvent(Typeahead.class, TypeaheadEvents.ACTIVE, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.IDLE, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.OPEN, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.CLOSE, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.CHANGE, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.RENDER, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.SELECT, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.AUTOCOMPLETE, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.CURSOR_CHANGE, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.ASYNC_REQUEST, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.ASYNC_CANCEL, ComponentCtrl.CE_IMPORTANT);
        addClientEvent(Typeahead.class, TypeaheadEvents.ASYNC_RECEIVE, ComponentCtrl.CE_IMPORTANT);
    }

    private Boolean highlight = true;
    private Boolean hint = true;
    private int minLength = 1;
    private String placeholder;
    private String dir;
    private Collection<Dataset> datasets = new ArrayList<>();

    public Boolean getHighlight() {
        return highlight;
    }

    public void setHighlight(Boolean highlight) {
        if (this.highlight != highlight) {
            this.highlight = highlight;
            smartUpdate("highlight", highlight);
        }
    }

    public boolean getHint() {
        return hint;
    }

    public void setHint(boolean hint) {
        if (this.hint != hint) {
            this.hint = hint;
            smartUpdate("hint", hint);
        }
    }

    public int getMinLength() {
        return minLength;
    }

    public void setMinLength(int minLength) {
        if (this.minLength != minLength) {
            this.minLength = minLength;
            smartUpdate("minLength", minLength);
        }
    }

    public String getPlaceholder() {
        return this.placeholder;
    }

    public void setPlaceholder(String placeholder) {
        if (this.placeholder == null || !this.placeholder.equals(placeholder)) {
            this.placeholder = placeholder;
            smartUpdate("placeholder", placeholder);
        }
    }

    public String getDir() {
        return this.placeholder;
    }

    public void setDir(String dir) {
        if (this.dir == null || !this.dir.equals(dir)) {
            this.dir = dir;
            smartUpdate("dir", dir);
        }
    }

    public Collection<Dataset> getDatasets() {
        return datasets;
    }


    public void setDataset(Dataset dataset) {
        this.datasets.clear();
        this.datasets.add(dataset);
        smartUpdate("datasets", datasets);
    }

    public void setDataset(Collection<Dataset> datasets) {
        this.datasets.clear();
        this.datasets.addAll(datasets);
        smartUpdate("datasets", datasets);
    }

    @Override
    protected void renderProperties(ContentRenderer renderer) throws IOException {
        super.renderProperties(renderer);
        render(renderer, "highlight", highlight);
        render(renderer, "hint", hint);
        render(renderer, "minLength", minLength);
        render(renderer, "placeholder", placeholder);
        render(renderer, "dir", dir);
        render(renderer, "datasets", datasets);
    }

}

File: src/main/resources/web/js/zul/typeahead/Typeahead.js
zul.typeahead.Typeahead = zk.$extends(zul.inp.Textbox, {

    // fields

    $define: {
        highlight: null,
        hint: null,
        minLength: null,
        placeholder: null,
        dir: null,
        dataset: null
    },

    // protected

    bind_: function () {
        this.$supers(zul.typeahead.Typeahead, 'bind_', arguments);
        var config = {
            hint: this.getHint(),
            highlight: this.getHighlight(),
            minLength: this.getMinLength()
        };
        var ds = this._getNativeDataset();
        var widget = this;
        var component = $('#' + this.uuid);
        component.typeahead(config, ds);
        component.bind('typeahead:active', function () {
            widget.fire('onTypeaheadActive');
        });
        component.bind('typeahead:idle', function () {
            widget.fire('onTypeaheadIdle');
        });
        component.bind('typeahead:open', function () {
            widget.fire('onTypeaheadOpen');
        });
        component.bind('typeahead:close', function () {
            widget.fire('onTypeaheadClose');
        });
        component.bind('typeahead:change', function (target, type, bubbles, cancelable) {
            widget.fire('onTypeaheadChange');
        });
        component.bind('typeahead:render', function (ev, suggestions, async, datasetName) {
            widget.fire('onTypeaheadRender', suggestions);
        });
        component.bind('typeahead:select', function (ev, suggestion) {
            widget.updateChange_();
            widget.fire('onSelect', suggestion);
        });
        component.bind('typeahead:autocomplete', function (ev, suggestion) {
            widget.fire('onTypeaheadAutocomplete', suggestion);
        });

        component.bind('typeahead:cursorchange', function (ev, suggestion) {
            widget.fire('onTypeaheadCursorChange', suggestion);
        });
        component.bind('typeahead:asyncrequest', function (ev, query, datasetName) {
            widget.fire('onTypeaheadAsyncRequest', query);
        });
        component.bind('typeahead:asynccancel', function (ev, query, datasetName) {
            widget.fire('onTypeaheadAsyncCancel', query);
        });
        component.bind('typeahead:asyncreceive', function (ev, query, datasetName) {
            widget.fire('onTypeaheadAsyncReceive', query);
        });
    },
    domClass_: function (no) {
        var classes = this.$supers("domClass_", no) || '';
        if (classes) {
            classes += ' ';
        }
        classes += 'z-textbox';
        return classes.trim();
    },

    // private

    _getNativeDataset: function () {
        var ds = this.getDataset();
        ds['templates'] = this._toNativeDatasetTemplates(ds['templates']);
        ds['source'] = this._toNativeDatasetSource(ds['source']);
        return ds;
    },
    _toNativeDatasetTemplates: function (templates) {
        for (var key in templates) {
            if (templates.hasOwnProperty(key)) {
                templates[key] = Handlebars.compile(templates[key]);
            }
        }
        return templates;
    },
    _toNativeDatasetSource: function (source) {
        if ('_class' in source) {
            switch (source['_class']) {
                case 'bloodhound':
                    if (source.hasOwnProperty('prefetch')) {
                        source.prefetch = this._toNativeFunctions(source.prefetch);
                    }
                    if (source.hasOwnProperty('remote')) {
                        source.remote = this._toNativeFunctions(source.remote);
                    }
                    return this._toBloodhoundDataset(source);
                default:
                    throw 'Unsupported dataset source: "' + source['_class'] + '"'
            }
        }
    },
    _toNativeFunctions: function (sourceData) {
        if (sourceData.hasOwnProperty('prepare')) {
            sourceData.prepare = this._toFunction(sourceData.prepare);
        }
        if (sourceData.hasOwnProperty('transform')) {
            sourceData.transform = this._toFunction(sourceData.transform);
        }
        return sourceData;
    },
    _toFunction: function (fun) {
        if (typeof fun == 'string') {
            return eval('(' + fun + ')');
        } else {
            return fun;
        }
    },
    _toBloodhoundDataset: function (source) {
        source['datumTokenizer'] = this._toBloodhoundTokenizers(source['datumTokenizers']);
        source['queryTokenizer'] = this._toBloodhoundTokenizers(source['queryTokenizers']);
        return new Bloodhound(source);
    },
    _toBloodhoundTokenizers: function (tokenizers) {
        if (tokenizers == undefined) {
            return null;
        } else if (tokenizers.length == 1) {
            return this._toBloodhoundTokenizer(tokenizers[0]);
        } else {
            var widget = this;
            var nativeTokenizers = $.map(tokenizers, function (tokenizer) {
                return widget._toBloodhoundTokenizer(tokenizer);
            });
            return function customTokenizer(datum) {
                return $.map(nativeTokenizers, function (nativeTokenizer) {
                    return nativeTokenizer(datum);
                });
            }
        }
    },
    _toBloodhoundTokenizer: function (tokenizer) {
        var hasKey = ('key' in tokenizer);
        var key = tokenizer['key'];
        var t = this._toBloodhoundRawTokenizer(tokenizer);
        if (!hasKey) {
            return t
        } else {
            return function (datum) {
                return t(datum[key]);
            }
        }
    },
    _toBloodhoundRawTokenizer: function (tokenizer) {
        switch (tokenizer['type']) {
            case 'nonword':
                return Bloodhound.tokenizers.nonword;
            case 'whitespace':
                return Bloodhound.tokenizers.whitespace;
            default:
                throw 'Unsupported Bloodhound tokenizer type: "' + tokenizer['type'] + '"'
        }
    }
});